posts/2022-11-03-Rust-1.65.0.md

```diff
@@ -0,0 +1,177 @@
+---
+layout: post
+title: "Announcing Rust 1.65.0"
+author: The Rust Release Team
+release: true
+---
+
+The Rust team is happy to announce a new version of Rust, 1.65.0. Rust is a
+programming language empowering everyone to build reliable and efficient
+software.
+
+---
+
+Before going into the details of the new Rust release, we'd like to draw
+attention to the tragic [death of Mahsa
+Amini](https://en.wikipedia.org/wiki/Death_of_Mahsa_Amini) and the death and
+violent suppression of many others, by the religious morality police of Iran.
+See <https://en.wikipedia.org/wiki/Mahsa_Amini_protests> for more details. We
+stand in solidarity with the people in Iran struggling for human rights.
+
+---
+
+If you have a previous version of Rust installed via rustup, you can get 1.65.0
+with:
+
+```console
+rustup update stable
+```
+
+If you don't have it already, you can [get
+`rustup`](https://www.rust-lang.org/install.html) from the appropriate page on
+our website, and check out the [detailed release notes for
+1.65.0](https://github.com/rust-lang/rust/blob/stable/RELEASES.md#version-1650-2022-11-03)
+on GitHub.
+
+If you'd like to help us out by testing future releases, you might consider
+updating locally to use the beta channel (`rustup default beta`) or the nightly
+channel (`rustup default nightly`). Please
+[report](https://github.com/rust-lang/rust/issues/new/choose) any bugs you
+might come across!
+
+## What's in 1.65.0 stable
+
+### Generic associated types (GATs)
+
+- [Stabilize generic associated types (GATs)](https://github.com/rust-lang/rust/pull/96709/)
+- Previous blog: <https://blog.rust-lang.org/2021/08/03/GATs-stabilization-push.html>
```

Comment:
```suggestion
- [More details in this recent blog post on Inside Rust](https://blog.rust-lang.org/2022/10/28/gats-stabilization.html)
- Previous blog: <https://blog.rust-lang.org/2021/08/03/GATs-stabilization-push.html>
```

Reply:
Oh, I only put these links for my own reference until we write something here for real. (The TODO just below.)

Reply:
Feel free to replace this as well, just wanted to make sure the link was handy. I do think the most recent blog post would be worth linking for more details.

Reply:
Here's a draft of this section (sorry it took so long!):

Lifetime, type, and const generics can now be defined on associated types, like so:

```rust
trait Foo {
    type Bar<'x>;
}
```

It's hard to put into few words just how useful these can be, so here are a
few example traits, to get a sense of their power:

```rust
/// An `Iterator`-like trait that can borrow from `Self`
trait LendingIterator {
    type Item<'a> where Self: 'a;

    fn next<'a>(&'a mut self) -> Option<Self::Item<'a>>;
}

/// Can be implemented over smart pointers, like `Rc` or `Arc`,
/// in order to allow being generic over the pointer type
trait PointerFamily {
    type Pointer<T>: Deref<Target = T>;

    fn new<T>(value: T) -> Self::Pointer<T>;
}

/// Allows borrowing an array of items. Useful for
/// `NdArray`-like types that don't necesarily store
/// data contiguously.
trait BorrowArray<T> {
    type Array<'x, const N: usize> where Self: 'x;

    fn borrow_array<'a, const N: usize>(&'a self) -> Self::Array<'a, N>;
}
```

As you can see, GATs are quite versatile and enable a number
of patterns that are not currently able to be written. For more
information, check out the post announcing the
[push for stabilization](https://blog.rust-lang.org/2021/08/03/GATs-stabilization-push.html) 
published last year or the
[stabilization announcement post](https://blog.rust-lang.org/2022/10/28/gats-stabilization.html) 
published last week. The former goes into a bit more depth
of a couple of the examples above, while the latter talks a
bit about some of the known limitations of this stabilization.

More in depth reading can be found in the associated types
section of the [nightly reference](https://doc.rust-lang.org/nightly/reference/items/associated-items.html#associated-types)
or the [original RFC](https://rust-lang.github.io/rfcs/1598-generic_associated_types.html) (which was initially opened over 6.5 years ago!).

Reply:
Added, thanks!